ykcompile/src/lib.rs

```diff
@@ -860,35 +460,35 @@ impl<TT> TraceCompiler<TT> {
             .collect::<Vec<u8>>();
 
         for arg in args {
+            // In which register will this argument be passed?
             // `unwrap()` must succeed, as we checked there are no more than 6 args above.
             let arg_reg = arg_regs.pop().unwrap();
 
-            match arg {
-                Operand::Place(place) => {
-                    // Load argument back from the stack.
-                    let (loc, _) = self.place_to_location(place, false);
-                    match &loc {
-                        Location::Register(reg) => {
-                            let off = stack_index(*reg) * 8;
-                            dynasm!(self.asm
-                                ; mov Rq(arg_reg), [rsp + off]
-                            );
-                        }
-                        Location::Mem(ro) => {
-                            dynasm!(self.asm
-                                ; mov Rq(arg_reg), [Rq(ro.reg) + ro.offs]
-                            );
-                        }
-                        Location::Addr(_) | Location::NotLive => unreachable!(),
-                    };
-                    self.free_if_temp(loc);
+            // Now load the argument into the correct argument register.
+            match self.iplace_to_location(arg) {
+                Location::Register(reg) => {
+                    // The value *was* in a register before we pushed it with caller_save().
+                    // We load it back from the stack now.
+                    // FIXME assumes that anything in a register was pushed. In reality only caller
+                    // save registers were. stack_index() will crash, so we will know when it's gone
+                    // wrong.
+                    let off = stack_index(reg) * 8;
+                    dynasm!(self.asm
+                        ; mov Rq(arg_reg), [rsp + off]
+                    );
                 }
-                Operand::Constant(c) => {
+                Location::Mem(ro) => dynasm!(self.asm
+                    ; mov Rq(arg_reg), [Rq(ro.reg) + ro.offs]
+                ),
+                Location::Indirect { .. } => todo!(),
+                Location::Const { val, .. } => {
+                    // FIXME assumes constant fits in a register.
```

Comment:
This sounds `assert`able?

Reply:
`i64_cast()` will crash if it doesn't fit.